public/scripts/reducers.js

```diff
@@ -25,14 +26,59 @@ function datasets(state = [], action) {
 }
 
 
-function featuresets(state={ featuresetList: [],
-                            featuresList: [] }, action) {
+function featuresets(state=[], action) {
   switch (action.type) {
     case Action.RECEIVE_FEATURESETS:
-      // {obs_features, sci_features}
-      return { ...state, featuresetList: action.payload };
-    case Action.RECEIVE_FEATURES:
-      return { ...state, features: action.payload };
+      return action.payload;
+    default:
+      return state;
+  }
+}
+
+
+function features(state={}, action) {
+  switch (action.type) {
+    case Action.RECEIVE_FEATURES: {
+      let tagList = [];
+      for (const feat in action.payload.tags) {
+        if (!action.payload.tags.hasOwnProperty(feat)) continue;
+        for (const idx in action.payload.tags[feat]) {
+          if (tagList.indexOf(action.payload.tags[feat][idx]) === -1) {
+            tagList.push(action.payload.tags[feat][idx]);
+          }
+        }
+      }
+      let allFeatsList = [];
+      for (const ctgy in action.payload.features_by_category) {
+        if (action.payload.features_by_category.hasOwnProperty(ctgy)) {
+          allFeatsList = allFeatsList.concat(action.payload.features_by_category[ctgy]);
+        }
+      }
+      return { ...action.payload,
+               allFeatsList,
+               tagList,
+               checkedTags: tagList.slice(0),
+               featsWithCheckedTags: allFeatsList.slice(0) };
+    }
+    case 'redux-form/CHANGE': {
```

Comment:
This is where we catch changes to the selected tags. If we were to have bound a callback to the `onChange` for each of the tag checkbox elements, it would've overridden the default `redux-form` action and the form state wouldn't have updated correctly.